src/main/java/org/jboss/remoting3/remote/ServerConnectionOpenListener.java

```diff
@@ -237,38 +237,51 @@ void initialiseCapabilities() {
 
 
 
+        @Override
         public void handleEvent(final ConnectedMessageChannel channel) {
             Pooled<ByteBuffer> pooledBuffer = connection.allocate();
             boolean free = true;
+            ByteBuffer receiveBuffer;
             try {
                 if (channel instanceof RemotingMessageChannel) {
-                    try {
-                        int messageLength = ((RemotingMessageChannel) channel).readMessageLength();
-                        if (messageLength > pooledBuffer.getResource().capacity() && messageLength < RemotingOptions.MAX_RECEIVE_BUFFER_SIZE) {
-                            pooledBuffer = Buffers.allocatedBufferPool(BufferAllocator.BYTE_BUFFER_ALLOCATOR, messageLength).allocate();
-                            ((RemotingMessageChannel) channel).adjustToMessageLength(messageLength);
+                    synchronized (connection.getLock()) {
+                        int res;
+                        RemotingMessageChannel.AdjustedBuffer ab = new RemotingMessageChannel.AdjustedBuffer(pooledBuffer);
+                        try {
+                            RemotingMessageChannel rc = (RemotingMessageChannel) channel;
+                            res = rc.receive(ab);
+                        } catch (IOException e) {
+                            connection.handleException(e);
+                            return;
+                        }
+                        if (res == -1) {
+                            log.trace("Received connection end-of-stream");
+                            connection.handlePreAuthCloseRequest();
+                            return;
+                        }
+                        if (res == 0) {
+                            return;
                         }
+                        pooledBuffer = ab.getAdjustedBuffer();
+                        receiveBuffer = pooledBuffer.getResource();
+                    }
+                } else {
+                    receiveBuffer = pooledBuffer.getResource();
+                    final int res;
```

Comment:
@bmaxwell should that be in synchronized block (https://github.com/jboss-remoting/jboss-remoting/blob/3.3/src/main/java/org/jboss/remoting3/remote/ServerConnectionOpenListener.java#L250) ?
It was added in 3.3 as part of REM3-182, but it looks relevant to this.